### PR TITLE
Characteristic functions for probability distributions

### PR DESCRIPTION
**Pull request by ethankward**
#### References to other Issues or PRs
Addresses #13833

#### Brief description of what is fixed or changed
Add function characteristic_function for all probability distributions (continuous, discrete and finite). Computed via integration / summation, which can be overridden in individual distribution implementations if necessary.

### CODE REVIEW

File: sympy/stats/rv.py [hunk 1]
```diff
@@ -765,6 +765,42 @@ def cdf(expr, condition=None, evaluate=True, **kwargs):
         return result
 
 
+def characteristic_function(expr, condition=None, evaluate=True, **kwargs):
+    """
+    Characterisfic Function of a random expression.
```
Comment (normalhuman) [inline sympy/stats/rv.py]: Typo in "Characteristic". I think Function should be lowercased here.

File: sympy/stats/rv.py [hunk 2]
```diff
@@ -765,6 +765,42 @@ def cdf(expr, condition=None, evaluate=True, **kwargs):
...
+
+    optionally given a second condition
```
Comment (normalhuman) [inline sympy/stats/rv.py]: This looks like a part of the previous sentence, it doesn't begin with an uppercase letter. Perhaps it should be moved up as a continuation of that sentence. (These docstrings go on display at docs.sympy.org where people don't expect uncapitalized sentence fragments).

File: sympy/stats/drv.py [hunk 3]
```diff
@@ -69,6 +69,21 @@ def cdf(self, x, **kwargs):
         """ Cumulative density function """
         return self.compute_cdf(**kwargs)(x)
 
+    @cacheit
+    def compute_characteristic_function(self, **kwargs):
+        """ Compute the characteristic function from the PDF
+
+        Returns a Lambda
+        """
+        x, t = symbols('x, t', integer=True, finite=True, cls=Dummy)
```
Comment (normalhuman) [inline sympy/stats/drv.py]: t is declared to be integer here but it should be real. Having it integer may lead to incorrect simplification of the sum, e.g., if the argument of `exp` ends up being an integer multiple of 2*pi*I for integer t.

File: sympy/stats/frv.py [hunk 4]
```diff
@@ -202,6 +202,11 @@ def pdf(self):
         return Lambda(x, Piecewise(*(
             [(v, Eq(k, x)) for k, v in self.dict.items()] + [(0, True)])))
 
+    @property
+    def characteristic_function(self):
+        t = Symbol('t')
+        return Lambda(t, sum(exp(I*k*t)*v for k,v in self.dict.items()))
```
Comment (normalhuman) [inline sympy/stats/frv.py]: In SymPy code, there is a space after comma

File: sympy/stats/frv.py [hunk 5]
```diff
@@ -202,6 +202,11 @@ def pdf(self):
         return Lambda(x, Piecewise(*(
             [(v, Eq(k, x)) for k, v in self.dict.items()] + [(0, True)])))
 
+    @property
+    def characteristic_function(self):
+        t = Symbol('t')
```
Comment (normalhuman) [inline sympy/stats/frv.py]: This should be a Dummy symbol (with real=True), as for other kinds of distributions. (The example in docstring also shows the inconsistency: sometimes _t, sometimes t. )

File: sympy/stats/crv.py [hunk 6]
```diff
@@ -398,6 +430,12 @@ def compute_cdf(self, expr, **kwargs):
         else:
             return ContinuousPSpace.compute_cdf(self, expr, **kwargs)
 
+    def compute_characteristic_function(self, expr, **kwargs):
+        if expr == self.value:
+            t = symbols("t", real=True, cls=Dummy)
+            return Lambda(t, self.distribution.characteristic_function(t, **kwargs))
+        else:return ContinuousPSpace.compute_characteristic_function(self, expr, **kwargs)
```
Comment (normalhuman) [inline sympy/stats/crv.py]: return should be on a new line, not right next to else: (PEP8)

**Review by normalhuman — CHANGES_REQUESTED** [review]
It's a good idea to add this method.

File: sympy/stats/rv.py (hunk 2 above)
Comment (ethankward) [reply sympy/stats/rv.py]: The same problem occurs in the cdf function above where I copied the docstring from, I'll fix my own but there should be a separate pull request for the other.

File: sympy/stats/rv.py [hunk 7]
```diff
@@ -765,6 +765,40 @@ def cdf(expr, condition=None, evaluate=True, **kwargs):
...
+    Examples
+    ========
+
+    >>> from sympy.stats import Normal, DiscreteUniform, Poisson, characteristic_function
+
+    >>> X = Normal('X', 0, 1)
+    >>> characteristic_function(X)
+    Lambda(_t, exp(-_t**2/2))
+
+    >>> Y = DiscreteUniform('Y', [1,2,7])
```
Comment (normalhuman) [inline sympy/stats/rv.py]: Spaces missing after commas in [1,2,7]

**Review by normalhuman — APPROVED** [review]
Apart from two missing spaces, looks good.

**Comment by jksuom** [comment]
@normalhuman Did you also check the expressions?

**Comment by asmeurer** [comment]
Can this be generalized higher up in the inheritance chain by simply using E[e^{itX}]?

**Comment by normalhuman** [comment]
@jksuom Yes, the three precomputed CFs (for exponential, normal, and uniform distributions) are correct.

@asmeurer Apparently not, because the classes SingleDiscreteDistribution, SingleContinuousDistribution, and SingleFiniteDistribution (where the `characteristic_function` method is added) do not inherit from a common class other than Basic and NamedArgsMixin. There is no SingleDistribution superclass for them.

**Comment by jksuom** [comment]
This looks good to me, thanks!

**Comment by normalhuman** [comment]
Thanks again, and a suggestion: it'd be better to create a branch for a feature instead of committing to master and then deleting the repo and forking again. Ref: [SymPy dev workflow](https://github.com/sympy/sympy/wiki/Development-workflow#create-a-new-branch)

**Pull request merged**